app/src/main/java/de/westnordost/streetcomplete/quests/construction/MarkCompletedHighwayConstruction.java

```diff
@@ -0,0 +1,119 @@
+package de.westnordost.streetcomplete.quests.construction;
+
+import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.text.TextUtils;
+import android.util.Log;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import de.westnordost.osmapi.map.data.BoundingBox;
+import de.westnordost.streetcomplete.R;
+import de.westnordost.streetcomplete.data.meta.OsmTaggings;
+import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
+import de.westnordost.streetcomplete.data.osm.download.MapDataWithGeometryHandler;
+import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
+import de.westnordost.streetcomplete.data.osm.tql.OverpassQLUtil;
+import de.westnordost.streetcomplete.quests.DateHandler;
+import de.westnordost.streetcomplete.quests.YesNoQuestAnswerFragment;
+
+public class MarkCompletedHighwayConstruction extends MarkCompletedConstruction
+{
+	@Inject
+	public MarkCompletedHighwayConstruction(OverpassMapDataDao overpassServer) {
+		super(overpassServer);
+	}
+
+	@Override public boolean download(BoundingBox bbox, MapDataWithGeometryHandler handler)
+	{
+		return overpassServer.getAndHandleQuota(getOverpassQuery(bbox), handler);
+	}
+
+	/** @return overpass query string to get streets marked as under construction but excluding ones
+	 * - with invalid construction tag
+	 * - with tagged opening date that is in future
+	 * - recently edited (includes adding/updating check_date tags)
+	 * - with fixme (as in some situations this quest may add fixme note as an edit)
+	 * . */
+	private static String getOverpassQuery(BoundingBox bbox)
+	{
+		ArrayList<String> acceptedConstructionValues = new ArrayList<String>(Arrays.asList(OsmTaggings.ALL_ROADS));
+		acceptedConstructionValues.add("cycleway");
+		acceptedConstructionValues.add("footway");
+		acceptedConstructionValues.add("path");
+
+		String groupName = ".roads_under_construction";
+		String resultsName = ".roads_for_review";
+		return OverpassQLUtil.getGlobalOverpassBBox(bbox) +
+			getQueryPart("highway", groupName, 14) +
+			"(" +
+			"way[construction]" +
+			"[construction !~ \"^("+ TextUtils.join("|", acceptedConstructionValues)+")$\"]" +
+			") -> .invalid_construction_type;" +
+			"(" +
+			"  way[highway=construction][fixme];" +
+			") -> .with_fixme;" +
+			"((" +
+			groupName + " - .invalid_construction_type)" +
+			"- .with_fixme" +
+			") -> " + resultsName + ";" +
+			resultsName + " out meta geom;";
```

Comment:
If the `getQueryPart` function would accept an additional parameter named something like `additionalFilters`, this would simplify this method somewhat

Reply:
I will try this.

Reply:
This code was already simplified by implementing other requests.

I am not convinced that this change would improve it further - it would just move complexity to getQueryPart function that would need to handle additionalFilters.

It would be even worse, as additionalFilters may be empty.